presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -49,34 +56,172 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator)
-                .forEach(symbol -> finalStatsBuilder.addSymbolStatistics(symbol, filteredStatistics.getSymbolStatistics(symbol)));
+        for (EquiJoinClause clause : criteria) {
+            SymbolStatsEstimate leftColumnStats = leftStats.getSymbolStatistics(clause.getLeft());
+            SymbolStatsEstimate rightColumnStats = rightStats.getSymbolStatistics(clause.getRight());
 
-        return Optional.of(finalStatsBuilder.build());
+            double leftNDV = leftColumnStats.getDistinctValuesCount();
+            double rightNDV = rightColumnStats.getDistinctValuesCount();
+
+            if (leftNDV > rightNDV) {
+                double selectedRangeFraction = leftColumnStats.getNonNullsFraction() * (leftNDV - rightNDV) / leftNDV;
+                double scaleFactor = selectedRangeFraction + leftColumnStats.getNullsFraction();
+                double newLeftNullsFraction = leftColumnStats.getNullsFraction() / scaleFactor;
+                outputStats = outputStats.mapSymbolColumnStatistics(clause.getLeft(), columnStats ->
+                        SymbolStatsEstimate.buildFrom(columnStats)
+                                .setNullsFraction(newLeftNullsFraction)
+                                .setDistinctValuesCount(leftNDV - rightNDV)
+                                .build());
+                outputStats = outputStats.mapOutputRowCount(rowCount -> rowCount * scaleFactor);
+            }
+            else if (leftNDV <= rightNDV) {
+                // only null values are left
+                outputStats = outputStats.mapSymbolColumnStatistics(clause.getLeft(), columnStats ->
+                        SymbolStatsEstimate.buildFrom(columnStats)
+                                .setNullsFraction(1.0)
+                                .setDistinctValuesCount(0.0)
+                                .build());
+                outputStats = outputStats.mapOutputRowCount(rowCount -> rowCount * leftColumnStats.getNullsFraction());
+            }
+            else {
+                // either leftNDV or rightNDV is NaN
+                return UNKNOWN_STATS;
+            }
+        }
+
+        return outputStats;
     }
 
-    private PlanNodeStatsEstimate crossJoinStats(JoinNode joinNode, Lookup lookup, Session session, Map<Symbol, Type> types)
+    @VisibleForTesting
+    PlanNodeStatsEstimate addAntiJoinStats(PlanNodeStatsEstimate joinStats, PlanNodeStatsEstimate antiJoinStats, Set<Symbol> joinSymbols)
     {
-        PlanNodeStatsEstimate leftStats = lookup.getStats(joinNode.getLeft(), session, types);
-        PlanNodeStatsEstimate rightStats = lookup.getStats(joinNode.getRight(), session, types);
+        checkState(joinStats.getSymbolsWithKnownStatistics().containsAll(antiJoinStats.getSymbolsWithKnownStatistics()));
+
+        double joinOutputRowCount = joinStats.getOutputRowCount();
+        double antiJoinOutputRowCount = antiJoinStats.getOutputRowCount();
+        double totalRowCount = joinOutputRowCount + antiJoinOutputRowCount;
```

Comment:
It's somehow unintuitive to use output stats before constructing them fully. I will leave as is